### PR TITLE
[WIP] Dynamic Commands (New PR for my old commands PR.)

### PR DESCRIPTION
**Pull request by AlphaModder**
Apparently github does not allow you to change the head fork after creating a PR, so i closed it and remade this one.

This slightly modifies the command interfaces to allow for more variety in command systems, and adds a more featured default command system The features it will introduce to it are:

* Allows use of /plugin:command in a less hacky way.
* Allows plugins to register conflicting commands, and resolves those conflicts when the command is sent.
* Removes the callback feature from command registration as it is no longer neccessary due to the above change.
* Allows use of /mc or /minecraft:command to run vanilla commands [NYI]
* Creates an API that may be used to set certain aliases to point to certain plugins in the command system.

It is based on EvilSeph's old blog post [here](http://evilseph.com/post/75608917830/bukkit-supporting-minecraft-commands).

Due to configuration files not yet having a standard format yet, I will only create an API that allows the command configuration described in the blog post to be implemented, but this PR does not include the actual implementation of the configuration file. 

I still need help with: a good way to wrap vanilla/sponge commands into the existing system. maybe some kind of VanillaPluginContainer/SpongePluginContainer as wrappers so the command system can use them?

### CODE REVIEW

File: src/main/java/org/spongepowered/api/plugin/Plugin.java [hunk 1]
```diff
@@ -62,19 +62,22 @@
     String version() default "unknown";
 
     /**
-     * A simple dependency string for this mod separated by a ";"
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/plugin/Plugin.java]: Please remove the reformatting

File: src/main/java/org/spongepowered/api/plugin/Plugin.java [hunk 2]
```diff
@@ -82,5 +85,9 @@
      */
     String dependencies() default "";
 
-
+    /**
+     * If this plugin is not purely a Sponge plugin. (Forge Mods, Vanilla
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/plugin/Plugin.java]: Start with a verb, e.g. `Checks`

File: src/main/java/org/spongepowered/api/plugin/Plugin.java [hunk 3]
```diff
@@ -82,5 +85,9 @@
...
+     * Command Wrapper...)
+     */
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/plugin/Plugin.java]: Missing `@return`

File: src/main/java/org/spongepowered/api/plugin/PluginContainer.java [hunk 4]
```diff
@@ -59,4 +59,12 @@
      */
     Object getInstance();
 
+    /**
+     * Gets if this PluginContainer wraps something that isn't a 'real' plugin.
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/plugin/PluginContainer.java]: What does 'real' mean? I would suggest 'loaded' or 'initialized' or something like that, since that's the big difference

**Comment by gratimax** [comment]
I think there's some better nomenclature we can have here. I suggest calling "wrapper" plugins "accessory".

File: src/main/java/org/spongepowered/api/service/command/CommandService.java [hunk 5]
```diff
@@ -25,83 +25,73 @@
...
      * @param aliases A list of aliases
-     * @return The registered command mapping, unless no aliases could be registered
-     * @throws IllegalArgumentException Thrown if {@code plugin} is not a plugin instance
+     * @return The registered command mapping, unless it could not be
+     *         registered.
+     * @throws IllegalArgumentException Thrown if {@code plugin} is not a plugin
+     *         instance
      */
-    Optional<CommandMapping> register(Object plugin, CommandCallable callable, List<String> aliases);
+    Optional<CommandMapping> register(Object plugin, CommandCallable callable, String... aliases);
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/service/command/CommandService.java]: Add back a way to register with `Iterable<String>`?

File: src/main/java/org/spongepowered/api/service/command/CommandService.java [hunk 6]
```diff
@@ -25,83 +25,73 @@
...
-     * @param alias An array of aliases
-     * @return The registered command mapping, unless no aliases could be registered
-     * @throws IllegalArgumentException Thrown if {@code plugin} is not a plugin instance
-     */
-    Optional<CommandMapping> register(Object plugin, CommandCallable callable, String... alias);
-
-    /**
-     * Register a given command using the given list of aliases.
-     *
-     * <p>If there is a conflict with one of the aliases (i.e. that alias
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/service/command/CommandService.java]: Why is this paragraph missing?

File: src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java [hunk 7]
```diff
@@ -64,9 +66,7 @@
  * {@link #onCommandEvent(CommandEvent)} method.
  */
 public class SimpleCommandService implements CommandService {
-
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java]: Formatting fixes

File: src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.spongepowered.api.service.command.sponge;
+
+import org.spongepowered.api.util.command.CommandSource;
+
+public interface AliasContext {
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java]: Missing javadocs

File: src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java [hunk 9]
```diff
@@ -360,4 +343,18 @@ public String getUsage() {
...
+        if (m.isPresent()) {
+            cmd.add(m.get());
+        }
+        return cmd;
+    }
+
+    @Override
+    public Optional<CommandMapping> resolveMapping(String alias, CommandSource source) {
+        return Optional.fromNullable(commands.get(alias.toLowerCase()));
+    }
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java]: Missing newline

**Comment by gratimax** [comment]
Pinging @sk89q to check if we want this kind of implementation in the API

**Comment by Zidane** [comment]
Cannot merge.

File: src/main/java/org/spongepowered/api/plugin/Plugin.java [hunk 10]
```diff
@@ -39,48 +39,54 @@
 public @interface Plugin {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/plugin/Plugin.java]: This entire file is filled with unnecessary edits. Please fix.

File: src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java [hunk 11]
```diff
@@ -231,8 +211,8 @@ public boolean call(CommandSource source, String arguments, List<String> parents
     }
 
     @Override
-    public boolean testPermission(CommandSource source) {
-        return dispatcher.testPermission(source);
+    public boolean testPermission(CommandSource source, boolean ignorePermissions) {
```
Comment (sk89q) [inline src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java]: Would an entirely new method not be more appropriate?

testPermissions(ignorePermissions = true) seems a bit roundabout.

**Comment by sk89q** [comment]
One thing we need... is for the command manager to handle success counts... since MC uses them with command blocks.

**Comment by AlphaModder** [comment]
@sk89q I can do that, if you like. :)

File: src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java (hunk 11 above)
Comment (AlphaModder) [inline src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java]: @sk89q Alright!

**Comment by jacklin213** [comment]
@AlphaModder You need to fix the licensing in both ```CommandResult``` and ```CommandResultBuilder```

File: src/main/java/org/spongepowered/api/plugin/PluginContainer.java (hunk 4 above)
Comment (me4502) [inline src/main/java/org/spongepowered/api/plugin/PluginContainer.java]: This may refer to the dummy field of the annotation, in which case it should say if it is a dummy plugin.

File: src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+package org.spongepowered.api.service.command.sponge;
+
+import org.spongepowered.api.util.command.CommandSource;
+
+/**
+ * Represents an alias that has been set to resolve to certain plugins under
+ * certain circumstances.
+ *
```
Comment (me4502) [inline src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java]: Redundant newline

File: src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
```
Comment (me4502) [inline src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java]: Redundant newline

File: src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandMapping.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
```
Comment (me4502) [inline src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandMapping.java]: Redundant newline

File: src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
```
Comment (me4502) [inline src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java]: Redundant newline

File: src/main/java/org/spongepowered/api/service/command/sponge/SpongeDispatcher.java [hunk 16]
```diff
@@ -0,0 +1,676 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
```
Comment (me4502) [inline src/main/java/org/spongepowered/api/service/command/sponge/SpongeDispatcher.java]: Redundant newline

File: src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java (hunk 13 above)
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java]: No, this is the correct formatting.
Google Java Style [Section 3](https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s3-source-file-structure) states that there is exactly one line between the license section and the package statement

Comment (Zidane) [inline src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java]: Simon we don't follow those 100%

File: src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java (hunk 9 above)
Comment (AlphaModder) [inline src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java]: @gratimax What? There's still one newline.

**Comment by simon816** [comment]
There's a bunch of unrelated changes made in this PR (e.g. import shuffling, javadoc wrapping etc.) which makes it difficult to see what has been added. Please fix this.

Also, there's merge conflicts that must be resolved.

File: src/main/java/org/spongepowered/api/event/SpongeEventFactory.java [hunk 17]
```diff
@@ -801,14 +804,16 @@ public static ProjectileLaunchEvent createProjectileLaunch(Game game, Cause caus
      * @param arguments The arguments provided
      * @param source The source of the command
      * @param command The command name
+     * @param result The result of the command
      * @return A new instance of the event
      */
-    public static CommandEvent createCommand(Game game, String arguments, CommandSource source, String command) {
+    public static CommandEvent createCommand(Game game, String arguments, CommandSource source, String command, Optional<CommandResult> result) {
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/event/SpongeEventFactory.java]: Don't use `Optional` in method parameters, instead, the parameter should be annotated with `@Nullable` and values.put should be `values.put("result", Optional.fromNullable(result));`

File: src/main/java/org/spongepowered/api/plugin/Plugin.java [hunk 18]
```diff
@@ -82,5 +82,11 @@
...
     String dependencies() default "";
 
-
+    /**
+     * Checks if this plugin is not purely a Sponge plugin. (Forge Mods, Vanilla
+     * Command Wrapper...)
+     * 
+     * @return If this plugin is not purely a Sponge plugin.
+     */
+    boolean dummy() default false;
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/plugin/Plugin.java]: Plugins should not set this, all `@Plugin`s will be a pure sponge plugin.

File: src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java [hunk 19]
```diff
@@ -104,75 +105,59 @@ public void onCommandEvent(final CommandEvent event) {
...
+        return register(plugin, callable, Arrays.asList(aliases), Functions.<List<String>>identity());
     }
 
     @Override
-    public Optional<CommandMapping> register(Object plugin, CommandCallable callable, List<String> aliases,
+    public Optional<CommandMapping> register(Object plugin, CommandCallable<Boolean> callable, List<String> aliases,
             Function<List<String>, List<String>> callback) {
         checkNotNull(plugin);
-
-        Optional<PluginContainer> containerOptional = this.pluginManager.fromInstance(plugin);
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java]: Don't remove `this.` from field/method accesses. In fact, the guidelines enforce the use of `this.` everywhere, you need to enforce it. (There's ways to do that in an IDE)

File: src/main/java/org/spongepowered/api/util/command/dispatcher/Dispatcher.java [hunk 20]
```diff
@@ -49,30 +54,33 @@
     /**
      * Get a list of primary aliases.
      *
-     * <p>The returned collection cannot be modified.</p>
+     * <p>
+     * The returned collection cannot be modified.
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/util/command/dispatcher/Dispatcher.java]: This is no longer a list/collection now it's a Map. The javadocs need updating and please explain what the keys/values represent.

File: src/main/java/org/spongepowered/api/service/command/sponge/CommandResult.java [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+     * @return If the command was processed
+     */
+    boolean wasProcessed();
+
+    /**
+     * Gets the success count of the command, if one exists.
+     * 
+     * @return The success count of the command
+     */
+    Optional<Integer> getSuccessCount();
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/command/sponge/CommandResult.java]: Do these need to be wrapped in `Optional` objects? Could it just return `0` rather than `Optional.absent()`?

File: src/main/java/org/spongepowered/api/util/command/dispatcher/Dispatcher.java [hunk 22]
```diff
@@ -49,30 +54,33 @@
...
 
     /**
-     * Get the {@link CommandCallable} associated with an alias. Returns
-     * null if no command is named by the given alias.
-     *
+     * Get the {@link CommandMapping}(s) associated with an alias.
+     * 
      * @param alias The alias
-     * @return The command mapping, if available
+     * @return The command mappings, if available
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/util/command/dispatcher/Dispatcher.java]: No need for 'if available' now it's not an `Optional`

File: src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java [hunk 23]
```diff
@@ -0,0 +1,257 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+public class SpongeCommandService implements CommandService<CommandResult> {
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java]: This is very similar to `SimpleCommandService`, how about making a base `AbscractCommandService` or something?

Comment (AlphaModder) [inline src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java]: Why? Especially considering they have different result types, it doesn't seem neccessary.

**Comment by boformer** [comment]
I think you misunderstand the ``parents`` argument of the ``CommandCallable``.

Here is an example what it does:
https://github.com/SpongePowered/SpongeAPI/issues/465#issuecomment-76290913

**Comment by boformer** [comment]
I also don't understand why you added Sponge-prefixed classes in a package named "org.spongepowered.api.service.command.sponge"

File: src/main/java/org/spongepowered/api/service/command/CommandService.java (hunk 6 above)
Comment (kenzierocks) [inline src/main/java/org/spongepowered/api/service/command/CommandService.java]: Because the whole command is?

**Comment by boformer** [comment]
Is the method ``CommandResult.getResultInfo()`` necessary to match the vanilla implementation?

**Comment by boformer** [comment]
I think we should change the class hierarchy. **It will simplify the command system dramatically.**

Right now the ``CommandService`` extends ``Dispatcher`` extends ``CommandCallable``.

That means that the ``CommandService`` inherits methods like ``getHelp``, ``getUsage`` or ``testPermission`` from the ``CommandCallable`` interface, which doesn't make sense (The service is not a command).

The reason why there is a separate ``Dispatcher`` interface that implements ``CommandCallable`` is to support subcommands. Plugins can create their own ``Dispatcher`` that can be registered as a command, but calls the registered subcommands (There is no documentation how it works, so it is not obvious at first).

**Comment by AlphaModder** [comment]
@boformer what else should I call them? 
And on the other thing, that's cool, but that's not really related to this PR, is it?

File: src/main/java/org/spongepowered/api/service/command/sponge/CommandResult.java (hunk 21 above)
Comment (ZephireNZ) [inline src/main/java/org/spongepowered/api/service/command/sponge/CommandResult.java]: Returning null means you can be certain there isn't a success count on the command.

**Comment by boformer** [comment]
Rename ``SpongeDispatcher`` to ``AliasContextDispatcher`` and move it to ``util.command.dispatcher``.
Same with other classes in the ``sponge`` package. Give them a proper name and move them to the right package.

I also don't understand why ``CommandCallable`` uses Generics. ``call`` should always return a ``CommandResult``.

I would also merge ``SpongeDispatcher`` and ``SpongeCommandService``, because ``SpongeDispatcher`` as a standalone class is not useful for plugin developers.

**Comment by AlphaModder** [comment]
@boformer, 
1. AliasContext/SpongeDispatcher is sponge dependent. sk89q has expressed his wish that the util package is not dependant on sponge. 
2. For CommandCallable, same reason. CommandResult is sponge-dependent. 
3. I agree.
4. Just to mimic the SimpleX's, I'm gonna keep it like this for now. 
5. That's not in the scope of this PR.

**Comment by boformer** [comment]
If it is a Sponge-dependent implementation, why don't you move the classes of the "sponge" package to the implementation?

If Sponge provides ``CommandService<CommandResult>``, all plugin commands have to be ``CommandCallable<CommandResult>``.

When a different implementation uses ``CommandService<Boolean>``, these plugins will not work with that implementation.

There has to be one standard.

**Comment by AlphaModder** [comment]
@boformer, because they should act the same in any implementation of Sponge, not just forge. Anything using the SpongeAPI should be based on it's command system, otherwise plugin commands would have to be different for each system.

**Comment by boformer** [comment]
Yes, that's why generics don't make sense.

**Comment by AlphaModder** [comment]
@boformer The commands API is not tied to SpongeAPI. It's sk89q's own thing. It doesn't always use CommandResult, only the SpongeAPI that uses it.

**Comment by boformer** [comment]
Sponge copied it from @sk89q, but that doesn't mean that we can not change it.

Generics just bloat up everything, and make it harder for plugin developers.

**Comment by AlphaModder** [comment]
@boformer I think it makes sense not to modify it. If you think it should be, you can make a PR after this one, and if others like it then it can be that way.

**Comment by boformer** [comment]
Well, looking at this again, there is no need for a "command" argument of the ``call`` method. The last element of the ``parents`` list is the command. I just misread the implementation code -_-

**Comment by gabizou** [comment]
@AlphaModder What do you want to do with this PR? It's been sitting for quite some time and no real decision from you has come from it as to its future.

**Comment by AlphaModder** [comment]
@gabizou It seems finished enough to me. I've been focusing on other PRs, but I can fix the merge conflicts. I don't know what else to add,

File: .gitignore [hunk 24]
```diff
@@ -64,6 +64,7 @@ Thumbs.db
 .idea
 .project
 .settings
+.metadata
```
Comment (kenzierocks) [inline .gitignore]: This seems largely irrelevant.

File: src/main/java/org/spongepowered/api/GameRegistry.java [hunk 25]
```diff
@@ -52,12 +51,15 @@
...
 import org.spongepowered.api.potion.PotionEffectBuilder;
 import org.spongepowered.api.potion.PotionEffectType;
+import org.spongepowered.api.service.command.sponge.CommandResultBuilder;
 import org.spongepowered.api.status.Favicon;
 import org.spongepowered.api.util.rotation.Rotation;
 import org.spongepowered.api.world.DimensionType;
 import org.spongepowered.api.world.biome.BiomeType;
 import org.spongepowered.api.world.difficulty.Difficulty;
 
+import com.google.common.base.Optional;
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/GameRegistry.java]: Wrong import order. This goes for all other files too.

File: src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java (hunk 19 above)
Comment (AlphaModder) [inline src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java]: @simon816 Sorry about that, it was just really annoying with the merge conflict.

**Comment by boformer** [comment]
I still don't understand why there are generic result types. ``CommandResult`` covers everything, so there is no need for an alternative ``Boolean`` result.

**Comment by gabizou** [comment]
@AlphaModder You have some decent PR's going on, but in every single one of them, you have formatting issues. I'd appreciate some courtesy in fixing your auto formatter since we do provide the formatting styles for both [IntelliJ](https://github.com/SpongePowered/SpongeAPI/blob/master/extra/intellij/sponge_intellij_style.xml) and [Eclipse](https://github.com/SpongePowered/SpongeAPI/tree/master/extra/eclipse).

I'd want to handle merging some PRs but I simply can't get over all the formatting issues. 

### FIX YOUR FORMATTING

**Comment by AlphaModder** [comment]
@Boformer CommandResult is for minecraft commands. boolean or otherwise is for other things utilizing the same commands API.

File: .gitignore (hunk 24 above)
Comment (gabizou) [inline .gitignore]: remove this please.

**Comment by Minecrell** [comment]
There are still too many formatting issues in here, makes it hard to review. Please use your IDE to add the qualified field accesses again and make sure you don't have changed empty lines everywhere.

**Comment by AlphaModder** [comment]
@Minecrell I can't find a way to qualify names in eclipse. Do you know one?

**Comment by Minecrell** [comment]
@AlphaModder The cleanup profile should do this for you.

**Comment by AlphaModder** [comment]
@Minecrell On it.

**Comment by AlphaModder** [comment]
@Minecrell Very strange, this is not adding thises when needed.

**Comment by Deamon5550** [comment]
@AlphaModder There are still some out of order imports such as [here](https://github.com/SpongePowered/SpongeAPI/pull/409/files#diff-7edfb10599c1944a8908a3af42d759bfL28) and [here](https://github.com/SpongePowered/SpongeAPI/pull/409/files#diff-d1733f2e4afa5036298e1bf8e37d122cL28).

Another thing is that you are removing a lot of empty lines uneccessarily for example from [here](https://github.com/SpongePowered/SpongeAPI/pull/409/files#diff-7edfb10599c1944a8908a3af42d759bfL28) and [also here](https://github.com/SpongePowered/SpongeAPI/pull/409/files#diff-63f5da161e3eeaf82738a30efdf3a625L131).

Additionally is the changing of the line wrap points in a lot of places, while technically correct it is out of scope for this PR. Some examples are [here](https://github.com/SpongePowered/SpongeAPI/pull/409/files#diff-7edfb10599c1944a8908a3af42d759bfL66) and [here](https://github.com/SpongePowered/SpongeAPI/pull/409/files#diff-d1733f2e4afa5036298e1bf8e37d122cL44).

I would suggest that you look through the entire diff of your changes and ask yourself if each line change is within the scope of your PR or if it is some unrelated or erroneous formatting change in which case it should be reverted. When we look through your PR we want to see exactly what is changing and not have to dig through a mountain of minor formatting modifications.

**Comment by AlphaModder** [comment]
Aaaaand the build failed for no reason. Amazing!

File: src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java [hunk 26]
```diff
@@ -140,15 +156,12 @@
                     throw new IllegalArgumentException("A command by the name of '" + alias + "' already exists");
                 }
             }
-
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java]: All these removed lines in here are out of scope for this PR

File: src/main/java/org/spongepowered/api/event/message/CommandEvent.java [hunk 27]
```diff
@@ -26,8 +26,12 @@
 package org.spongepowered.api.event.message;
 
 import org.spongepowered.api.event.GameEvent;
+import org.spongepowered.api.service.command.sponge.CommandResult;
 import org.spongepowered.api.util.command.CommandSource;
 import org.spongepowered.api.util.event.Cancellable;
+import org.spongepowered.api.util.event.Order;
+
+import com.google.common.base.Optional;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/event/message/CommandEvent.java]: Wrong import order.

File: src/main/java/org/spongepowered/api/event/message/CommandEvent.java [hunk 28]
```diff
@@ -61,4 +65,13 @@
...
 
+    /**
+     * The result of the command. This is only available when the event is being
+     * handled with {@link Order#LATE}.
+     * 
+     * @return The result of the command, if present
+     */
+    Optional<CommandResult> getResult();
+
+
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/event/message/CommandEvent.java]: Extra line.

File: src/main/java/org/spongepowered/api/service/command/CommandService.java [hunk 29]
```diff
@@ -36,72 +36,66 @@
...
-     * @throws IllegalArgumentException Thrown if {@code plugin} is not a plugin instance
+     * @return The registered command mapping, unless it could not be
+     *         registered.
+     * @throws IllegalArgumentException Thrown if new conflicting aliases are
+     *         added in the callback
+     * @throws IllegalArgumentException Thrown if {@code plugin} is not a plugin
+     *         instance
      */
-    Optional<CommandMapping> register(Object plugin, CommandCallable callable, List<String> aliases, Function<List<String>, List<String>> callback);
+    public Optional<CommandMapping> register(Object plugin, CommandCallable<T> callable, List<String> aliases,
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/service/command/CommandService.java]: Remove the `public` modifier since it is redundant.

File: src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java [hunk 30]
```diff
@@ -0,0 +1,255 @@
...
+        return register(plugin, callable, Arrays.asList(aliases));
+    }
+
+    @Override
+    public Optional<CommandMapping> register(Object plugin, CommandCallable<CommandResult> callable, List<String> aliases,
+            Function<List<String>, List<String>> callback) {
+        return register(plugin, callable, aliases);
+    }
+
+    public Optional<CommandMapping> register(Object plugin, CommandCallable<CommandResult> callable, List<String> aliases) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java]: Is this method overridden or a unique method of `SpongeCommandService`?

File: src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java [hunk 31]
```diff
@@ -0,0 +1,255 @@
...
+    @Override
+    public Optional<CommandMapping> resolveMapping(String alias, CommandSource source) {
+        return this.dispatcher.resolveMapping(alias, source);
+    }
+
+    public void addAliasContext(String alias, AliasContext context) {
+        this.dispatcher.addAliasContext(alias, context);
+    }
+
+    public void removeAliasContext(String alias) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java]: Likewise here about overrides.

File: src/main/java/org/spongepowered/api/util/command/CommandCallable.java [hunk 32]
```diff
@@ -33,40 +33,60 @@
...
  *
- * <p>Implementations are not required to implement a sane
- * {@link java.lang.Object#equals(Object)} but may choose to do so.</p>
+ * <p>
+ * Implementations are not required to implement a sane
+ * {@link java.lang.Object#equals(Object)} but may choose to do so.
+ * </p>
  */
-public interface CommandCallable extends CommandCompleter {
+public interface CommandCallable<T> extends CommandCompleter {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/util/command/CommandCallable.java]: You'll likely need javadocs for `@param <T>`

File: src/main/java/org/spongepowered/api/util/command/CommandMapping.java [hunk 33]
```diff
@@ -45,22 +47,25 @@
...
+     * although generally implementations should ignore the casing of aliases.
+     * </p>
      *
      * @return A set of aliases
      */
     Set<String> getAllAliases();
 
     /**
-     * Gets the callable.
+     * Get the callable
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/util/command/CommandMapping.java]: The original doc was correct....

File: src/main/java/org/spongepowered/api/util/command/dispatcher/Dispatcher.java [hunk 34]
```diff
@@ -28,51 +28,59 @@
...
+import org.spongepowered.api.util.command.CommandSource;
 
+import java.util.Map;
 import java.util.Set;
 
 /**
  * Executes a command based on user input.
  */
-public interface Dispatcher extends CommandCallable {
+public interface Dispatcher<T> extends CommandCallable<T> {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/util/command/dispatcher/Dispatcher.java]: Likewise here, you'll need javadocs for `@param <T>`

File: src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java [hunk 35]
```diff
@@ -22,7 +22,6 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  * THE SOFTWARE.
  */
-
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java]: Unnecessary line change.

File: src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java [hunk 36]
```diff
@@ -269,13 +266,11 @@ public synchronized boolean containsAlias(String alias) {
     @Override
     public boolean containsMapping(CommandMapping mapping) {
         checkNotNull(mapping);
-
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java]: Unnecessary line change.

File: src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java [hunk 37]
```diff
@@ -289,17 +284,14 @@ public synchronized int size() {
...
+    public Boolean call(CommandSource source, String arguments, List<String> parents) throws CommandException {
         String[] parts = arguments.split(" +", 2);
-        Optional<CommandMapping> mapping = get(parts[0]);
-
+        Optional<CommandMapping> mapping = resolveMapping(parts[0], source);
         if (mapping.isPresent()) {
             List<String> passedParents = new ArrayList<String>(parents.size() + 1);
             passedParents.addAll(parents);
             passedParents.add(parts[0]);
-
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java]: Unnecessary line change.

File: src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java (hunk 31 above)
Comment (AlphaModder) [inline src/main/java/org/spongepowered/api/service/command/sponge/SpongeCommandService.java]: @gabizou These are not overrides.

File: src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java [hunk 38]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * Gets if this alias context should be used when running from the specified
+     * command source.
+     * 
+     * @param source The command source to check.
+     * @return If this alias context should be used when running from the
+     *         specified command source.
+     */
+    public boolean appliesTo(CommandSource source);
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/command/sponge/AliasContext.java]: `public` is redundant here

**Comment by gabizou** [comment]
Shifting this to another PR for continued discussion and refactoring.
